servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Collectors.java

```diff
@@ -0,0 +1,170 @@
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.CompletableSource.Subscriber;
+import io.servicetalk.concurrent.internal.DuplicateSubscribeException;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.internal.SubscriberUtils.handleExceptionFromOnSubscribe;
+import static java.util.concurrent.atomic.AtomicIntegerFieldUpdater.newUpdater;
+
+/**
+ * A factory for creating a {@link Collector} factory and an optional buffer boundaries {@link Publisher} for
+ * typical buffering needs.
+ */
+final class Collectors {
+
+    private Collectors() {
+        // no instances
+    }
+
+    static <T> CollectorFactory<T> forSizeOrDuration(final int bufferSize, final Duration duration,
+                                                     final Executor executor) {
+        SizeBasedSignal sizeBasedSignal = new SizeBasedSignal(bufferSize);
+        return new DefaultCollectorFactory<>(sizeBasedSignal,
+                sizeBasedSignal.ambWith(executor.timer(duration).beforeOnComplete(sizeBasedSignal::reset))
+                        .repeat(__ -> true));
+    }
+
+    /**
+     * A {@link Supplier} of {@link Collector}s and a {@link Publisher} of buffer boundaries.
+     *
+     * @param <T> Type of items collected.
+     */
+    interface CollectorFactory<T> extends Supplier<Collector<T, Iterable<T>>> {
+        Publisher<?> boundaries();
+    }
+
+    private static final class DefaultCollectorFactory<T> implements CollectorFactory<T> {
+        private final SizeBasedSignal sizeBasedSignal;
+        private final Publisher<?> boundaries;
+
+        DefaultCollectorFactory(final SizeBasedSignal sizeBasedSignal, final Publisher<?> boundaries) {
+            this.sizeBasedSignal = sizeBasedSignal;
+            this.boundaries = boundaries;
+        }
+
+        @Override
+        public Publisher<?> boundaries() {
+            return boundaries;
+        }
+
+        @Override
+        public Collector<T, Iterable<T>> get() {
+            return new ListCollector<>(sizeBasedSignal);
+        }
+    }
+
+    private static final class ListCollector<T> implements Collector<T, Iterable<T>> {
+        private final List<T> buffer = new ArrayList<>();
+        private final SizeBasedSignal sizeBasedSignal;
+
+        ListCollector(final SizeBasedSignal sizeBasedSignal) {
+            this.sizeBasedSignal = sizeBasedSignal;
+        }
+
+        @Override
+        public void add(@Nullable final T t) {
+            try {
+                buffer.add(t);
+            } finally {
+                sizeBasedSignal.itemAdded();
+            }
+        }
+
+        @Override
+        public Iterable<T> close() {
```

Comment:
nit: I think we should throw in `add` after `close` was called. This should be part of the interface contract imho